include/oneapi/tbb/detail/_flow_graph_impl.h

```diff
@@ -359,21 +361,23 @@ class graph : no_copy, public graph_proxy {
 
 };  // class graph
 
+template<typename DerivedType>
 inline void graph_task::destruct_and_deallocate(const execution_data& ed) {
     auto allocator = my_allocator;
     // TODO: investigate if direct call of derived destructor gives any benefits.
     this->~graph_task();
-    allocator.deallocate(this, ed);
+    allocator.deallocate(static_cast<DerivedType*>(this), ed);
```

Comment:
As far as I understand, the root cause of the issue is the automatic upcast of the `this` pointer to `graph_task` during the call to function template `allocator.deallocate()`.

Can we introduce an overload to `deallocate` function such as:
```cpp
    void deallocate(void* ptr, size_t size, const execution_data& ed);
```
to have:
```suggestion
    allocator.deallocate(this, sizeof(this), ed);
```
instead? 

This way allows us avoiding introduction of one more template parameter.

Reply:
You cannot remove `static_cast` to the derived type because it is not guaranteed (in a general case) that the derived object has the same address as the parent object.

As for one more templated parameter, `sizeof(this)` in the parent class code is `sizeof` of the base class, in other words, `sizeof(this) != sizeof(static_cast<Derived*>(this))`. (C++ does not have reflection (yet?), i.e. we cannot get size information of the real object from the pointer to the parent object)

Reply:
Yeah. I agree.